src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts

```diff
@@ -246,6 +261,138 @@ export class InstructorCoursesPageComponent implements OnInit {
+    return courseModel;
+  }
+
+  /**
+   * Copies a feedback session.
+   */
+  private copyFeedbackSession(fromFeedbackSession: FeedbackSession, newSessionName: string, newCourseId: string):
+      Observable<FeedbackSession> {
+    let createdFeedbackSession!: FeedbackSession;
+    return this.feedbackSessionsService.createFeedbackSession(newCourseId, {
+      feedbackSessionName: newSessionName,
+      instructions: fromFeedbackSession.instructions,
+
+      submissionStartTimestamp: fromFeedbackSession.submissionStartTimestamp,
+      submissionEndTimestamp: fromFeedbackSession.submissionEndTimestamp,
+      gracePeriod: fromFeedbackSession.gracePeriod,
+
+      sessionVisibleSetting: fromFeedbackSession.sessionVisibleSetting,
+      customSessionVisibleTimestamp: fromFeedbackSession.customSessionVisibleTimestamp,
+
+      responseVisibleSetting: fromFeedbackSession.responseVisibleSetting,
+      customResponseVisibleTimestamp: fromFeedbackSession.customResponseVisibleTimestamp,
+
+      isClosingEmailEnabled: fromFeedbackSession.isClosingEmailEnabled,
+      isPublishedEmailEnabled: fromFeedbackSession.isPublishedEmailEnabled,
+    }).pipe(
+      switchMap((feedbackSession: FeedbackSession) => {
+        createdFeedbackSession = feedbackSession;
+        return this.feedbackQuestionsService.getFeedbackQuestions({
+          courseId: fromFeedbackSession.courseId,
+          feedbackSessionName: fromFeedbackSession.feedbackSessionName,
+          intent: Intent.FULL_DETAIL,
+        });
+      }),
+      switchMap((response: FeedbackQuestions) => {
+        if (response.questions.length === 0) {
+          // no questions to copy
+          return of(createdFeedbackSession);
+        }
+        return from(response.questions).pipe(
+          concatMap((feedbackQuestion: FeedbackQuestion) => {
+            return this.feedbackQuestionsService.createFeedbackQuestion(
+              createdFeedbackSession.courseId, createdFeedbackSession.feedbackSessionName, {
+                questionNumber: feedbackQuestion.questionNumber,
+                questionBrief: feedbackQuestion.questionBrief,
+                questionDescription: feedbackQuestion.questionDescription,
+
+                questionDetails: feedbackQuestion.questionDetails,
+                questionType: feedbackQuestion.questionType,
+
+                giverType: feedbackQuestion.giverType,
+                recipientType: feedbackQuestion.recipientType,
+
+                numberOfEntitiesToGiveFeedbackToSetting: feedbackQuestion.numberOfEntitiesToGiveFeedbackToSetting,
+                customNumberOfEntitiesToGiveFeedbackTo: feedbackQuestion.customNumberOfEntitiesToGiveFeedbackTo,
+
+                showResponsesTo: feedbackQuestion.showResponsesTo,
+                showGiverNameTo: feedbackQuestion.showGiverNameTo,
+                showRecipientNameTo: feedbackQuestion.showRecipientNameTo,
+              });
```

Comment:
Likewise, any possibility to abstract away this?
Otherwise, it is fine to leave as is; just somewhat unnecessarily chunky here reading as "copying everything from `feedbackQuestion` to a new struct that inherits all the field values => semantically equivalent to a `clone()`

Reply:
ok! will split this into a separate function too